src/lib.rs

```diff
@@ -284,35 +269,71 @@ impl<R: gimli::Reader> Context<R> {
         &self.sections
     }
 
-    fn find_unit_id(&self, probe: u64) -> Option<usize> {
-        let idx = self.unit_ranges.binary_search_by(|r| {
-            if probe < r.0.begin {
-                Ordering::Greater
-            } else if probe >= r.0.end {
-                Ordering::Less
-            } else {
-                Ordering::Equal
-            }
-        });
-        let idx = match idx {
-            Ok(x) => x,
-            Err(_) => return None,
+    // Finds the CGU id (index in `self.units`) for the function address given.
+    // The index of the address in the CGU's address table is also returned
+    // because this is calculated here to ensure the function address actually
+    // resides in the functions of the CGU.
+    fn find_unit_id_and_address(&self, probe: u64) -> Option<(usize, usize)> {
+        // First up find the first position in the array which could have our
+        // function address.
+        let pos = match self.unit_ranges.binary_search_by_key(&probe, |r| r.0.begin) {
+            // Although unlikely, we could find an exact match.
+            Ok(i) => i,
+            // `probe` is smaller than all CGU start addresses, so it's not
+            // contained anywhere.
+            Err(0) => return None,
+            // No exact match was found, but this probe would fit at slot `i`.
+            // This means that slot `i-1` is the highest start address which is
+            // underneath `probe`, which is the CGU we'd like to start
+            // searching.
```

Comment:
Nit: DWARF calls them compilation units (CU).

Given that CUs can be much larger than needed, it's not clear to me that this is the CU we need to start at. Why can't there be a CU with a lower start address that contains the probe address?

Reply:
Hm yes I'm entirely mixed up in how sorting works, let me see if I can find a solution which actually works.

Reply:
Ok I think the new logic should work out!